src/zcash/address/zip32.cpp

```diff
@@ -153,13 +297,73 @@ SaplingExtendedFullViewingKey SaplingExtendedSpendingKey::ToXFVK() const
     return ret;
 }
 
-libzcash::SaplingPaymentAddress SaplingExtendedSpendingKey::DefaultAddress() const
-{
-    return ToXFVK().DefaultAddress();
+//
+// Unified
+//
+
+std::optional<std::pair<UnifiedSpendingKey, HDKeyPath>> UnifiedSpendingKey::ForAccount(const HDSeed& seed, uint32_t bip44CoinType, uint32_t accountId) {
+    UnifiedSpendingKey usk;
+    usk.accountId = accountId;
+
+    auto transparentKey = DeriveZip32TransparentAccountKey(seed, bip44CoinType, accountId);
+    if (!transparentKey.has_value()) return std::nullopt;
+    usk.transparentKey = transparentKey.value().first;
+
+    auto saplingKey = SaplingExtendedSpendingKey::ForAccount(seed, bip44CoinType, accountId);
+    usk.saplingKey = saplingKey.first;
+
+    return std::make_pair(usk, saplingKey.second);
+}
+
+UnifiedFullViewingKey UnifiedSpendingKey::ToFullViewingKey() const {
+    UnifiedFullViewingKey ufvk;
+
+    if (transparentKey.has_value()) {
+        ufvk.transparentKey = transparentKey.value().Neuter();
+    }
+
+    if (saplingKey.has_value()) {
+        ufvk.saplingKey = saplingKey.value().ToXFVK();
+    }
+
+    return ufvk;
+}
+
+std::optional<ZcashdUnifiedAddress> UnifiedFullViewingKey::Address(diversifier_index_t j) const {
+    ZcashdUnifiedAddress ua;
+
+    if (transparentKey.has_value()) {
+        // ensure that the diversifier index is small enough for a t-addr
+        if (MAX_TRANSPARENT_CHILD_IDX < j) return std::nullopt;
+
+        CExtPubKey changeKey;
+        if (!transparentKey.value().Derive(changeKey, 0)) {
+            return std::nullopt;
+        }
+
+        CExtPubKey childKey;
+        unsigned int childIndex = (unsigned int) j.GetUint64(0);
+        if (changeKey.Derive(childKey, childIndex)) {
+            ua.transparentKey = childKey.pubkey;
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    if (saplingKey.has_value()) {
+        auto saplingAddress = saplingKey.value().Address(j);
+        if (saplingAddress.has_value()) {
+            ua.saplingAddress = saplingAddress.value();
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    return ua;
 }
 
 std::optional<unsigned long> ParseZip32KeypathAccount(const std::string& keyPath) {
-    std::regex pattern("m/32'/[0-9]+'/([0-9]+)'");
+    std::regex pattern("m/32'/[0-9]+'/([0-9]+)'.*");
```

Comment:
I don't see much of a point to exactly match the child levels? I guess it could exclude some malformed paths. However, this should probably be made generic over the `purpose` field since we're now going to have to cope with having both `44'` and `32'` as purpose strings.

Reply:
The function as-is remains accurate because it is defined as the ZIP 32 keypath account, but I agree that we should just have a generic function that knows how to find the account level from any keypath we support.